### PR TITLE
Add possibility for custom naming when using multiple dataloaders

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/6091
Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/6245

Allows user to disable the automatic addtion `/dataloader_idx_{idx}` at the end of logged values. This allows users to give custom names depending on dataset for example:
``` python
if dataloader_idx == 0:
    `self.log('val_loss_cifar', val, auto_add_dataloader_idx=False)
else:
    `self.log('val_loss_mnist', val, auto_add_dataloader_idx=False)
```
## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Review by s-rog — APPROVED** [review]

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -226,6 +226,7 @@ def log(
         sync_dist: bool = False,
         sync_dist_op: Union[Any, str] = 'mean',
         sync_dist_group: Optional[Any] = None,
+        auto_add_dataloader_idx: bool = True,
```
Comment (carmocca) [inline pytorch_lightning/core/lightning.py:229]:
```suggestion
        add_dataloader_idx: bool = True,
```

File: tests/trainer/logging_/test_logger_connector.py [hunk 2]
```diff
@@ -470,3 +471,37 @@ def training_step(self, *args, **kwargs):
     )
     with pytest.warns(UserWarning, match="The progress bar already tracks a metric with the .* 'loss'"):
         trainer.fit(model)
+
+
+@pytest.mark.parametrize("auto_add_dataloader_idx", [False, True])
+def test_auto_add_dataloader_idx(tmpdir, auto_add_dataloader_idx):
+    """ test that auto_add_dataloader_idx argument works """
+
+    class TestModel(EvalModelTemplate):
```
Comment (carmocca) [inline tests/trainer/logging_/test_logger_connector.py:480]: Can you update the test to use BoringModel instead?

**Comment by tchaton** [comment]
Hey @SkafteNicki,

I though we should have done something slightly different for this one.
We could use `dataset.__name__` to resolve it automatically or find a way for the users to return the dataset 

Best,
T.C

**Review by tchaton — APPROVED** [review]
LGTM !

**Comment by SkafteNicki** [comment]
> Hey @SkafteNicki,
> 
> I think we could do something slightly different for this one.
> We could use `dataset.__name__` or find a way for the users to provide the dataset names directly and automate this for the users.
> 
> Best,
> T.C

@tchaton if you look at the two issues, one user request that they can append the `dataloader_idx` in front and the second user ask for appending the dataset name. I do not think we can find an automatic way for accommodating everyone, therefore I just propose to give the user a way to disable the current behaviour and give them full control over the naming.

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (SkafteNicki) [reply pytorch_lightning/core/lightning.py]: changed :]

File: tests/trainer/logging_/test_logger_connector.py (hunk 2 above)
Comment (SkafteNicki) [reply tests/trainer/logging_/test_logger_connector.py]: Done :]

**Review by carmocca — APPROVED** [review]

**Comment by george-gca** [comment]
I am with @SkafteNicki on this one. I tend to use the name as `dataset/metric` so it is correctly logged in tensorboard. Also sometimes I want to save the metric for a specific test image, so I generate it as `dataset/image_name/metric`.
